### PR TITLE
Convert to Hetionet v1.0 identifiers

### LINKED ISSUE #12: Hetnet IDs for Disease/Compound IDs
Currently Pubtator uses MeSH ids for disease and compound mentions, but for our interests we will need to convert these ids into the same ids hetnets are using. @dhimmel let me know if you want to grab the list or I can if that is desired.

### PR DESCRIPTION
**Pull request by danich1**
This pr closes #12 . The idea behind this pull request is to incorporate hetnet ids, when extracting pubtator tags. Hetnet ids will only appear if there the MeSH ids match the id mapper provided in the repository. Assigning review to @dhimmel and @zietzm, since you both have worked on this repository before. Let me know what you guys think.

### CODE REVIEW

**Comment by dhimmel** [comment]
Let's focus this pull request just on the sample data, i.e. generating a converted [`data/example/3-sample-tags.tsv`](https://github.com/greenelab/pubtator/blob/1d99f00afddff65931218ca68961abe24d398983/data/example/3-sample-tags.tsv). Then once that is working, we can scale to all of pubtator.

We'll only focus on the `Disease`, `Compound`, and `Gene` types, so other annotation types will be filtered out.

File: scripts/pubtator_to_xml.py [hunk 1]
```diff
@@ -134,36 +183,81 @@ def read_bioconcepts2pubtator_offsets(path):
...
+    Return a dictionary of key - disease terms value - hetnet id
+    """
+    with open("../disease_hetnet_ids.tsv") as id_file:
+        id_reader = csv.DictReader(id_file, delimiter="\t")
+        disease_map = {row["resource_id"]: row["doid_code"] for row in id_reader}
+    return disease_map
+
+
+def clean_up_annotations(lines):
+    """
```
Comment (danich1) [inline scripts/pubtator_to_xml.py]: quoting=csv.QUOTE_NONE

File: data/example/4-hetnet-tags.tsv [hunk 2]
```diff
@@ -0,0 +1,22 @@
+Document	Type	ID	Offset	End
```
Comment (dhimmel) [inline data/example/4-hetnet-tags.tsv]: Nice!

If it doesn't break too much stuff, I'd consider renaming columns like: `pubmed_id`, `kind`, `identifier`, `offset`, `end`.

File: environment.yml [hunk 3]
```diff
@@ -5,3 +5,4 @@ dependencies:
 - pip:
     - git+https://github.com/en-dash/PyBioC@a21bd181ce3b89de791db4bb7db6ca1c48167648
     - tqdm==4.11.2
+    - pandas=0.20.2
```
Comment (dhimmel) [inline environment.yml]: Use pandas from anaconda (not pip).

File: mapper/map_to_drugbank.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+drug_map_df[drug_map_df.ID_TYPE == 'DRUGBANK_ID'][['DRUG_ID', 'IDENTIFIER']].rename(columns={'IDENTIFIER': 'drugbank_id'})
+).drop('DRUG_ID', axis='columns')
+
+# Grab the chebi to DB mapper
+url = 'https://raw.githubusercontent.com/dhimmel/drugbank/7b94454b14a2fa4bb9387cb3b4b9924619cfbd3e/data/mapping/chebi.tsv'
+chebi_map_df = pd.read_table(url)
+chebi_map_df["chebi_id"] = chebi_map_df["chebi_id"].astype(object)
+
+merged_df = pd.merge(chebi_map_df, drug_map_df, on="drugbank_id", how="outer")
+merged_df.to_csv("drugbank_mapper.tsv", sep="\t", index=False)
```
Comment (dhimmel) [inline mapper/map_to_drugbank.py]: Add final newline

File: scripts/hetnet_id_extractor.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    final_df[["Document", "Type", "ID", "Offset", "End"]].to_csv(args.output, sep="\t", index=False)
+
+
+def load_disease_df():
+    """
+    Load the mesh-hetnet id mapping table.
+    Return a dataframe of id mappings
+    """
+    url = 'https://raw.githubusercontent.com/dhimmel/disease-ontology/052ffcc960f5897a0575f5feff904ca84b7d2c1d/data/xrefs-prop-slim.tsv'
+    return pd.read_csv(url, sep="\t")
```
Comment (dhimmel) [inline scripts/hetnet_id_extractor.py]: Use read_table which defaults to TSV

File: mapper/drugbank_mapper.tsv [hunk 6]
```diff
@@ -0,0 +1,8926 @@
+drugbank_id	chebi_id	MeSH
+DB00035	4450	D003894
```
Comment (dhimmel) [inline mapper/drugbank_mapper.tsv]: I think we want this file to have only two columns: `drugbank_id` and `mesh_id`. No duplicate rows, no missing values, and sorted.

File: tests.sh [hunk 7]
```diff
@@ -9,3 +9,8 @@ python scripts/pubtator_to_xml.py \
 python scripts/extract_tags.py \
   --input data/example/2-sample-docs.xml \
   --output data/example/3-sample-tags.tsv
+
+# Filter-Convert tags to Hetnet IDs
+python scripts/hetnet_id_extractor.py \
+  --input data/example/3-sample-tags.tsv \
+  --output data/example/4-hetnet-tags.tsv
```
Comment (dhimmel) [inline tests.sh]: add final newline.

File: mapper/map_to_drugbank.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
+import pandas as pd
+
+# Map from DrugBank to MeSH using DrugCentral
+url = 'https://github.com/olegursu/drugtarget/blob/9a6d84bed8650c6c507a2d3d786814c774568610/identifiers.tsv?raw=true'
+drug_map_df = pd.read_table(url)
+drug_map_df = drug_map_df[drug_map_df.ID_TYPE.str.contains('MESH')][['DRUG_ID', 'IDENTIFIER']].rename(columns={'IDENTIFIER': 'MeSH'}).merge(
```
Comment (dhimmel) [inline mapper/map_to_drugbank.py]: There's a more modern way to format this:

```python
(drug_map_df
    [drug_map_df.ID_TYPE.str.contains('MESH')]
    [['DRUG_ID', 'IDENTIFIER']]
    .rename(columns={'IDENTIFIER': 'MeSH'})
    .merge(
        drug_map_df
        [drug_map_df.ID_TYPE == 'DRUGBANK_ID']
        [['DRUG_ID', 'IDENTIFIER']]
        .rename(columns={'IDENTIFIER': 'drugbank_id'})
    ).drop('DRUG_ID', axis='columns')
```

File: environment.yml [hunk 9]
```diff
@@ -5,3 +5,4 @@ dependencies:
 - pip:
     - git+https://github.com/en-dash/PyBioC@a21bd181ce3b89de791db4bb7db6ca1c48167648
     - tqdm==4.11.2
```
Comment (dhimmel) [inline environment.yml]: tqdm is also now in anaconda channel... so move pandas and tqdm above.

File: data/example/4-hetnet-tags.tsv [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+100001	Compound	9606	22	30
+100002	Compound	9606	92	100
+100002	Compound	9606	1026	1034
+100002	Compound	9606	1297	1305
+100002	Compound	9606	1363	1370
+100003	Compound	9606	83	88
+100013	Compound	D013311	276	290
+100014	Compound	D009638	836	850
+100014	Compound	D009638	1173	1187
+100001	Disease		36	60
```
Comment (dhimmel) [inline data/example/4-hetnet-tags.tsv]: Looks like disease IDs are mising?

File: scripts/hetnet_id_extractor.py [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+    final_df[["pubmed_id", "type", "identifier", "offset", "end"]].to_csv(args.output, sep="\t", index=False)
+
+
+def load_disease_df():
+    """
+    Load the mesh-hetnet id mapping table.
+    Return a dataframe of id mappings
+    """
+    url = 'https://raw.githubusercontent.com/dhimmel/disease-ontology/052ffcc960f5897a0575f5feff904ca84b7d2c1d/data/xrefs-prop-slim.tsv'
```
Comment (dhimmel) [inline scripts/hetnet_id_extractor.py]: Do we want to filter this dataframe to just the relevant types for pubtator here?

Also can this file be untracked from this repo, if we're reading from URL?

File: environment.yml [hunk 12]
```diff
@@ -2,6 +2,7 @@ name: pubtator
 dependencies:
 - python=3.6.0
 - lxml=3.7.2
+- tqdm=4.11.2
+- pandas=0.20.2
 - pip:
-    - git+https://github.com/en-dash/PyBioC@a21bd181ce3b89de791db4bb7db6ca1c48167648
-    - tqdm==4.11.2
+    - git+https://github.com/en-dash/PyBioC@a21bd181ce3b89de791db4bb7db6ca1c48167648
```
Comment (dhimmel) [inline environment.yml]: Add final newline. Note that all files tracked by git should end with a final newline. This helps with diffing the file going forward. See the red symbol github shows at the end of this line due to missing newline.

File: scripts/hetnet_id_extractor.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
+import argparse
+
+import pandas as pd
+
+
+def filter_tags(infile, outfile):
+    extracted_tag_df = pd.read_csv(infile, sep="\t")
```
Comment (dhimmel) [inline scripts/hetnet_id_extractor.py]: `read_table`

File: scripts/hetnet_id_extractor.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    """
+    return pd.read_table("mapper/drugbank_mapper.tsv")
+
+
+def load_gene_df():
+    """
+    Loads the gene table from hetnets
+    Return a dataframe with the gene ids
+    """
+    url = 'https://raw.githubusercontent.com/dhimmel/entrez-gene/a7362748a34211e5df6f2d185bb3246279760546/data/genes-human.tsv'
```
Comment (dhimmel) [inline scripts/hetnet_id_extractor.py]: Must subset to protein-coding genes.

File: scripts/hetnet_id_extractor.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+    # Verify Gene IDs are human genes
+    gene_df = extracted_tag_df[extracted_tag_df["type"] == "Gene"]
+    gene_final_df = gene_df[gene_df["identifier"].isin(hetnet_gene_df["GeneID"])]
+
+    final_df = gene_final_df
+    final_df = final_df.append(chemical_merged_df)
+    final_df = final_df.append(disease_merged_df)
+
+    final_df[["pubmed_id", "type", "identifier", "offset", "end"]].to_csv(args.output, sep="\t", index=False)
```
Comment (dhimmel) [inline scripts/hetnet_id_extractor.py]: I think we still want the output sorted by pubmed_id rather than type?

File: environment.yml [hunk 16]
```diff
@@ -2,6 +2,8 @@ name: pubtator
 dependencies:
 - python=3.6.0
 - lxml=3.7.2
+- tqdm=4.11.2
+- pandas=0.20.2
 - pip:
     - git+https://github.com/en-dash/PyBioC@a21bd181ce3b89de791db4bb7db6ca1c48167648
-    - tqdm==4.11.2
+
```
Comment (dhimmel) [inline environment.yml]: The problem here (why github shows the red symbol) is that the final newline has some whitespace on it. Instead it should be blank.

File: scripts/hetnet_id_extractor.py [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+
+
+def load_disease_df():
+    """
+    Load the mesh-hetnet id mapping table.
+    Return a dataframe of id mappings
+    """
+    url = 'https://raw.githubusercontent.com/dhimmel/disease-ontology/052ffcc960f5897a0575f5feff904ca84b7d2c1d/data/xrefs-prop-slim.tsv'
+    disease_df = pd.read_table(url)
+    return disease_df[disease_df["resource"] == "MSH"]
```
Comment (dhimmel) [inline scripts/hetnet_id_extractor.py]: Do we also need to include OMIM terms?

Comment (danich1) [inline scripts/hetnet_id_extractor.py]: No they are only using mesh ids from what I saw in the annotated file.

File: tests.sh [hunk 18]
```diff
@@ -9,3 +9,9 @@ python scripts/pubtator_to_xml.py \
...
+
```
Comment (dhimmel) [inline tests.sh]: Whitespace issue here.

**Pull request merged**